crates/matrix-sdk-ui/src/notification_client.rs

```diff
@@ -84,94 +94,317 @@ impl NotificationClient {
+            (StateEventType::RoomMember, "$LAZY".to_owned()),
+            (StateEventType::RoomMember, "$ME".to_owned()),
+            (StateEventType::RoomCanonicalAlias, "".to_owned()),
+            (StateEventType::RoomName, "".to_owned()),
+            (StateEventType::RoomPowerLevels, "".to_owned()), /* necessary to build the push
+                                                               * context */
+        ];
+
+        let invites = SlidingSyncList::builder("invites")
+            .sync_mode(SlidingSyncMode::new_selective().add_range(0..=16))
+            .timeline_limit(8)
+            .required_state(required_state.clone())
+            .filters(Some(assign!(SyncRequestListFilters::default(), {
+                is_invite: Some(true),
+                is_tombstoned: Some(false),
+                not_room_types: vec!["m.space".to_owned()],
+            })))
+            .sort(vec!["by_recency".to_owned(), "by_name".to_owned()]);
+
+        let sync = self
+            .client
+            .sliding_sync(Self::CONNECTION_ID)?
+            .poll_timeout(Duration::from_secs(1))
+            .network_timeout(Duration::from_secs(1))
+            .with_account_data_extension(
+                assign!(AccountDataConfig::default(), { enabled: Some(true) }),
+            )
+            .add_list(invites)
+            .build()
+            .await?;
+
+        sync.subscribe_to_room(
+            room_id.to_owned(),
+            Some(assign!(RoomSubscription::default(), {
+                required_state,
+                timeline_limit: Some(uint!(16))
+            })),
+        );
+
+        let mut remaining_attempts = 3;
+
+        let stream = sync.sync();
+        pin_mut!(stream);
+
+        loop {
+            if stream.next().await.is_none() {
+                // Sliding sync aborted early.
+                break;
+            }
 
-        let (sender_display_name, sender_avatar_url) = match sender {
-            Some(sender) => (
-                sender.display_name().map(|s| s.to_owned()),
-                sender.avatar_url().map(|s| s.to_string()),
-            ),
-            None => (None, None),
+            if notification.lock().unwrap().is_some() {
+                // We got the event.
+                break;
+            }
+
+            remaining_attempts -= 1;
+            if remaining_attempts == 0 {
+                // We're out of luck.
+                break;
+            }
+        }
+
+        self.client.remove_event_handler(stripped_member_handler);
+        self.client.remove_event_handler(timeline_event_handler);
+
+        let maybe_event = notification.lock().unwrap().take();
+        Ok(maybe_event)
+    }
+
+    /// Get a full notification, given a room id and event id.
+    ///
+    /// This will run a small sliding sync to retrieve the content of the event,
+    /// along with extra data to form a rich notification context.
+    ///
+    /// For this to work, this sliding sync must not interfere with other
```

Comment:
Doesn't the notification client need to use an in-memory store even if it's in the same process? Can't we enforce that the `NotificationClient` uses the memory store? I think we should be able to do so, no?

Reply:
That's now done, by inheriting the parent `Client`'s configuration into one that uses an in-memory state store in place of the one configured by the parent :+1: